packages/app/components/jsonschemaform/widgets/utils.ts

```diff
@@ -0,0 +1,47 @@
+import { ID_PREFIX } from './constants'
+
+const macros = {
+    FIELD_LOOKUP: (name: string) => {
+        let el = document.getElementById(`${ID_PREFIX}${name}`) as HTMLInputElement
+        // This may be run before the element has been rendered to the DOM, so don't return `undefined`.
+        const value = el.value ?? ''
+
+        return globalThis.encodeURIComponent(value)
+    },
+} as const
+
+function findAndReplace(
+    sourceString: string = '',
+    every: string = '',
+    to: string[] = [],
+    macros: Record<string, Function> = {}
+) {
+    // Starting from the source string, iterate through every entry in "to", replacing with either a macro's return value or a value.
+    return to.reduce((accumulator, current) => {
+        // We may have a MACRO_NAME:FieldName combo, or we may have a value.
+        const [macroNameOrValue, maybeFieldName] = current.split(':')
```

Comment:
I didn't write an ABFN grammar this time for our little macro DSL, but the idea is commended here. The tests show that you can pass in any combo: a value, a MACRO:value, or a MACRO, and this will still work.